src/librustc_mir/interpret/place.rs

```diff
@@ -1,472 +1,755 @@
+    mplace: MemPlace,
+    pub layout: TyLayout<'tcx>,
+}
+
+impl<'tcx> ::std::ops::Deref for MPlaceTy<'tcx> {
+    type Target = MemPlace;
+    fn deref(&self) -> &MemPlace {
+        &self.mplace
     }
+}
+
+impl<'tcx> From<MPlaceTy<'tcx>> for PlaceTy<'tcx> {
+    fn from(mplace: MPlaceTy<'tcx>) -> Self {
+        PlaceTy {
+            place: Place::Ptr(mplace.mplace),
+            layout: mplace.layout
+        }
+    }
+}
 
-    pub fn from_scalar_ptr(ptr: ScalarMaybeUndef, align: Align) -> Self {
-        Place::Ptr {
+impl MemPlace {
+    #[inline(always)]
+    pub fn from_scalar_ptr(ptr: Scalar, align: Align) -> Self {
+        MemPlace {
             ptr,
             align,
             extra: PlaceExtra::None,
         }
     }
 
+    #[inline(always)]
+    pub fn from_ptr(ptr: Pointer, align: Align) -> Self {
+        Self::from_scalar_ptr(ptr.into(), align)
+    }
+
+    #[inline(always)]
+    pub fn to_scalar_ptr_align(self) -> (Scalar, Align) {
+        assert_eq!(self.extra, PlaceExtra::None);
+        (self.ptr, self.align)
+    }
+
+    /// Extract the ptr part of the mplace
+    #[inline(always)]
+    pub fn to_ptr(self) -> EvalResult<'tcx, Pointer> {
+        // At this point, we forget about the alignment information -- the place has been turned into a reference,
+        // and no matter where it came from, it now must be aligned.
+        self.to_scalar_ptr_align().0.to_ptr()
+    }
+
+    /// Turn a mplace into a (thin or fat) pointer, as a reference, pointing to the same space.
+    /// This is the inverse of `ref_to_mplace`.
+    pub fn to_ref(self, cx: impl HasDataLayout) -> Value {
+        // We ignore the alignment of the place here -- special handling for packed structs ends
+        // at the `&` operator.
+        match self.extra {
+            PlaceExtra::None => Value::Scalar(self.ptr.into()),
+            PlaceExtra::Length(len) => Value::new_slice(self.ptr.into(), len, cx),
+            PlaceExtra::Vtable(vtable) => Value::new_dyn_trait(self.ptr.into(), vtable),
+        }
+    }
+}
+
+impl<'tcx> MPlaceTy<'tcx> {
+    #[inline]
+    fn from_aligned_ptr(ptr: Pointer, layout: TyLayout<'tcx>) -> Self {
+        MPlaceTy { mplace: MemPlace::from_ptr(ptr, layout.align), layout }
+    }
+
+    #[inline]
+    pub(super) fn len(self) -> u64 {
+        // Sanity check
+        let ty_len = match self.layout.fields {
+            layout::FieldPlacement::Array { count, .. } => count,
+            _ => bug!("Length for non-array layout {:?} requested", self.layout),
+        };
+        if let PlaceExtra::Length(len) = self.extra {
+            len
+        } else {
+            ty_len
+        }
+    }
+}
+
+// Validation needs to hash MPlaceTy, but we cannot hash Layout -- so we just hash the type
+impl<'tcx> Hash for MPlaceTy<'tcx> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.mplace.hash(state);
+        self.layout.ty.hash(state);
+    }
+}
+impl<'tcx> PartialEq for MPlaceTy<'tcx> {
```

Comment:
can't this be derived?

Reply:
No, `TyLayout` has no `PartialEq`. So we just compare the type.

Reply:
You might want to include a few more things in the Eq/Hash, such as an `Option<usize>` where `Some(index)` means `Variants::Single { index }`, while `None` is for all other `Variants`.